File: java_benchmarks/src/main/java/com/cmpe220/benchmark/CassandraWriteUnstructured.java
package com.cmpe220.benchmark;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Base64;

import com.datastax.oss.driver.api.core.CqlSession;
import com.datastax.oss.driver.api.core.cql.*;

/**
* REQUIRED: have Cassandra DB already running
*/
public class CassandraWriteUnstructured extends AbstractBenchmark {

    private CqlSession session;
    private String img_blob;

    public CassandraWriteUnstructured() {
        category = "cassandra";
        description = "This is an example";
    }

    public void setupQuery() throws IOException {
        System.out.println("Connecting to cassandra...");
        session = CqlSession.builder().build();

        session.execute(
            "CREATE KEYSPACE cmpe220KS "+
            "WITH replication="+
            "{'class':'SimpleStrategy','replication_factor':1}"
        );
        session.execute("USE cmpe220KS");
        session.execute("CREATE TABLE tbl (col1 INT PRIMARY KEY, col2 MAP<TEXT, BLOB>)");
        img_blob = new String(
            Base64.getEncoder().encode(
                Files.readAllBytes(Paths.get("data/cat.JPG"))
            )
        );
    }

    public void endQuery() {
        System.out.println("Closing...");
        session.execute("DROP KEYSPACE cmpe220KS");
        session.close();
    }

    public void runQuery() {
        session.execute(
            String.format(
                "INSERT INTO tbl (col1, col2) VALUES(1, {'cat': textAsBlob('%s')})", img_blob
            )
        );
    }
}

File: python_benchmarks/CassandraWriteUnstructured.py
'''
REQUIRED: have Cassandra DB already running
pip install cassandra-driver (https://docs.datastax.com/en/developer/python-driver/3.25/)

(This is run from the parent directory, so
"python_benchmarks.AbstractBenchmark"
is required instead of "AbstractBenchmark")
'''
from python_benchmarks.AbstractBenchmark import AbstractBenchmark
from cassandra.cluster import Cluster
import base64


class CassandraWriteUnstructured(AbstractBenchmark):

    def __init__(self):
        self.category = 'cassandra'
        self.description = 'This is an example'

    def setupQuery(self):
        print('Connecting to cassandra...')
        cluster = Cluster()
        self.session = cluster.connect()

        self.session.execute(
            "CREATE KEYSPACE cmpe220KS "
            "WITH replication="
            "{'class':'SimpleStrategy','replication_factor':1}"
        )
        self.session.execute("USE cmpe220KS")
        self.session.execute('CREATE TABLE tbl'
                             '(col1 INT PRIMARY KEY, col2 MAP<TEXT, BLOB>)')
        with open('data/cat.JPG', 'rb') as img:
            self.img_blob = base64.b64encode(img.read()).decode('utf-8')

    def endQuery(self):
        print('Closing...')
        self.session.execute("DROP KEYSPACE cmpe220KS")

    def runQuery(self):
        self.session.execute(
            "INSERT INTO tbl (col1, col2) VALUES(1, {'cat': textAsBlob('%s')})" % self.img_blob
        )
